lib/use.spec.ts

```diff
@@ -4,4 +4,130 @@ describe("use", () => {
   it("throws an error for invalid middleware", () => {
     expect(() => use(() => null)).toThrowError();
   });
+
+  it("resolves with the middleware response if middleware bypasses handler", async () => {
+    const apiHandler = () => {
+      return Promise.resolve('handler response!')
+    }
+
+    const middleware = async (req, res, next) => {
+      await next();
+      return Promise.resolve('middleware response!');
+    }
+
+    const wrappedHandler = use(middleware)(apiHandler)
+
+    const result = await wrappedHandler();
+
+    expect(result).toEqual('middleware response!');
```

Comment:
Hey Ev, 

Could you share a bit more about the intended use case for this behavior? I only just noticed that your examples are demonstrating a wrapped handler that provides a return value.

Returning values within Next.js API requests isn't a pattern I have encountered before. Although you are correct that this makes unit tests a bit more complex, I'm not sure that I see a decent use case for this in production.

The intention for `use` and `label` is to be placed nearest the default export in a given Next.js API route. Invoking them inside of an API handler wasn't a pattern I intended on supporting. 

Do you think your use case (aborting a request and returning an error) would be satisfied with a pattern like this?

```ts
import type { NextApiHandler } from "next"; 
import { use } from "next-api-middleware";

const handler: NextApiHandler = async (req, res) => {
  // Never gets called
  res.status(200);
  res.json({
    foo: "bar";
  });
};

export default use(
  async (req, res, next) => {
    try {
      // Let subsequent middleware execute
      await next();
    } catch (err) {
      // Capture any errors and respond with an HTTP 500
      res.status(500);
      res.json({
        error: err.message,
      });
    }
  },
  async (req, res, next) => {
    // Throw an error to abort the chain
    throw new Error("Something broke");
  }
)(handler);
```

Reply:
Sure, the use-case is as follows. I'm building a fairly large app using Next.js which includes a REST API with many endpoints. Those endpoints share a lot of the same basic requirements (authentication, input validation, error handling, etc...). `next-api-middleware` is fantastic for allowing me to abstract all those common bits into little utils that I can use to wrap my API methods.

To illustrate this in code, here is an example of a couple middleware functions I have (simplified for brevity):

```ts
// Authentication middleware
const authentication: Middleware = async (req, res, next) => {
    const isLoggedIn = await myAuthModule.check(req);
    if (!isLoggedIn) return res.status(401).json({errors: ['Not authenticated.']})
    return next();
}

// Input validation middleware
const inputValidation: Middleware = async (req, res, next) => {
    const isValid = await yup.object().shape({test: yup.string()}).validate(req.body);
    if (!isValid) return res.status(422).json({errors: ['Invalid input.']})
    return next();
}
```

I then wrap my API routes with these middleware methods via `use`:

```ts
# /pages/api/myendpoint.ts

const handler = () => {...}

export default use(authentication, inputValidation)(handler);
```

As you can see each middleware layer does whatever check it needs to make, aborts with the appropriate error code if there's a failure, otherwise continues down the chain. Next, I have unit tests which test all my endpoints to ensure they behave appropriately:

```ts
import endpoint from './pages/api/myendpoint';

it('should throw a 401 error if user is not authenticated', async () => {
	const {req, res} = mockRequestResponse();

	await endpoint(req, res);

	expect(res._getJSONData()).toEqual({
		errors: ['Not authenticated.'],
	});
	expect(res._getStatusCode()).toBe(401);
});

it('should throw a 422 error if invalid data given', async () => {
        // Mock the authentication module to ensure we're logged in
        mock(myAuthModule, 'check').mockResolvedValue(true)
       
	const {req, res} = mockRequestResponse();
	
        // Mock invalid endpoint payload
        req.body = JSON.stringify({foo: 'invalid'})

	await endpoint(req, res);

	expect(res._getJSONData()).toEqual({
		errors: ['Invalid input.'],
	});
	expect(res._getStatusCode()).toBe(422);
});
```

With my PR this is all possible and works great because each middleware does not allow the next to be executed unless it passes. It allows each middleware to abort the promise chain cleanly and it allows my tests to `await` the endpoint and catch the error so that I can assert the response.

> The intention for `use` and `label` is to be placed nearest the default export in a given Next.js API route. Invoking them inside of an API handler wasn't a pattern I intended on supporting.

Not sure what you mean here. I'm using `use` to wrap my API routes as designed I think. I'm not invoking the middleware directly.

> Do you think your use case (aborting a request and returning an error) would be satisfied with a pattern like this?

I don't think so. In your example there's only 1 global error returned if something goes wrong (the 500 error). I want my middleware to abort the chain and return the specialized error for whatever that middleware is checking for.

> Returning values within Next.js API requests isn't a pattern I have encountered before.

That's strange. I wonder if there's a different way to do it then. In my opinion, what I have is perfectly reasonable. It works just like any regular Promise chain. Any part of the chain can throw an error and that will abort the rest of the chain. I can then attach a `.then()` and listen to whatever the first error in the chain was.

My PR allows this behavior. You can have both worlds. You can either have your middleware abort the chain and return that response, or if you can let it fallthrough to the next middleware.

Reply:
Given the explanation above... I personally think it is a nice addition to `next-api-middleware` package, especially, if it is a non-breaking change. In other words, if you can still use the package as originally designed and have to make absolutely no changes if you do not wish to have this use case, then it's all beneficial.

Perhaps the PR could include `documentation` on `when` do you want to use this approach, `why` what is better/worse than existing approaches, `what` does it actually do, `how` do you do it.

Just my 2cents. I am not an owner. :p

Reply:
Good idea @FranciscoKloganB. I added a short paragraph to the README. I didn't want to go into lengthy docs until I hear more from @htunnicliff on the status of this PR. Happy to write more examples and docs if there's still interest in merging this in.